### DISCUSSION TITLE
How to properly type-check SQLA v1.4 code (and avoid mypy crashes)

### DISCUSSION THREAD

**Discussion by jenstroeger** [sqlalchemy/sqlalchemy#10950] [Usage Questions] [answered]
Hello, Using SQLA v1.4 and Python 3.10 with type hints, running `mypy` v1.8.0 crashes. So do all versions back to v1.4.0, so does tip-of-tree `mypy` as of today and commit [c3cd83a](https://github.com/python/mypy/tree/c3cd83a23dcdbd72e6e2a4c2d6c605eadc7bccf4).

There are a few related discussions, e.g.

- https://github.com/sqlalchemy/sqlalchemy/discussions/9014
- https://github.com/sqlalchemy/sqlalchemy/issues/10282

Currently I’m running on
```
mypy                               1.8.0
SQLAlchemy                         1.4.51
sqlalchemy2-stubs                  0.0.2a38
```
and what I’m seeing is
```
/path/to/project/src/orm/mappings.py:68: error: INTERNAL ERROR -- Please try using mypy master on GitHub:
https://mypy.readthedocs.io/en/stable/common_issues.html#using-a-development-mypy-build
Please report a bug at https://github.com/python/mypy/issues
version: 1.9.0+dev.c3cd83a23dcdbd72e6e2a4c2d6c605eadc7bccf4
Traceback (most recent call last):
  File "/path/to/project/.venv/bin/mypy", line 8, in <module>
    sys.exit(console_entry())
  ...
  File "/path/to/project/.venv/lib/python3.10/site-packages/mypy/semanal.py", line 1816, in apply_class_plugin_hooks
    hook(ClassDefContext(defn, base_expr, self))
  File "/path/to/project/.venv/lib/python3.10/site-packages/sqlalchemy/ext/mypy/plugin.py", line 267, in _base_cls_hook
    decl_class.scan_declarative_assignments_and_apply_types(ctx.cls, ctx.api)
  File "/path/to/project/.venv/lib/python3.10/site-packages/sqlalchemy/ext/mypy/decl_class.py", line 102, in scan_declarative_assignments_and_apply_types
    _scan_declarative_assignment_stmt(
  File "/path/to/project/.venv/lib/python3.10/site-packages/sqlalchemy/ext/mypy/decl_class.py", line 390, in _scan_declarative_assignment_stmt
    assert node is lvalue.node
AssertionError: 
/path/to/project/src/orm/mappings.py:68: : note: use --pdb to drop into pdb
```
However, as @zzzeek states [here](https://github.com/sqlalchemy/sqlalchemy/issues/10282#issuecomment-1693755241):

> we unfortunately don't have resources to keep tracking these down as mypy keeps changing but we do accept pull requests with tests.

which leaves me unsure about how to proceed from here…

- Play with `mypy` options?
- Create a [minimal, reproducible example](https://stackoverflow.com/help/minimal-reproducible-example)?
- Debug the extension (much new code to learn)?
- Try another checker, e.g. [pyright](https://github.com/microsoft/pyright) but that [doesn’t support SQLA’s dynamic features](https://github.com/microsoft/pyright/issues/626).

I noticed that _removing_ the `sqlalchemy2-stubs` package avoids `mypy` crashing but it leaves me with errors like
```
/path/to/some_file.py: error: Skipping analyzing
"sqlalchemy.orm.attributes": module is installed, but missing library stubs or
py.typed marker  [import-untyped]
```
but it also delivers some useful output and errors.

What are your recommendations to proceed here?

**Comment by CaselIT** [comment, marked as answer]
Hi,

I would suggest
- upgrade to sqlalchemy 2. Using typing in previous version is not great and prone to frustrations
- use a mypy version that does not crashes
- try to debug the mypy plugin and create a PR. If it doesn't break the current test PR for the mypy plugin are still accepted
- try to create a reproducible example and open an issue, but I sqlalchemy team can't promise it will be fixed


regarding pyright, in v2 most things should work with it too.

**Reply by jenstroeger** [reply]
Thanks @CaselIT, I think we’ll have to bite the bullet and move to SQLA v2. Sooner or later that has to happen anyway…

**Comment by jenstroeger** [comment]
@CaselIT as I’m reviewing code and migrate to SQLA v2, I run into type issues like
```
error: Call to untyped function "create" in typed context  [no-untyped-call]
```
for Postgres `DOMAIN` types, because it seems that the parent class [`NamedType`](https://github.com/sqlalchemy/sqlalchemy/blob/a8dbf8763a8fa2ca53cc01033f06681a421bf60b/lib/sqlalchemy/dialects/postgresql/named_types.py#L31) isn’t fully typed. I can silence that error for now, but what’s the recommendation here? Open an issue? Try to add types and open a PR?

**Reply by CaselIT** [reply]
Not everything it still fully typed, even in v2. Most of the public api should though. For the rest you can refer to the list here https://github.com/sqlalchemy/sqlalchemy/issues/6810 (that may not be 100% up to date)

For that particular module I don't remember is it was typed in one of the PR yet to merge (I'll check), but in general if you have time to contribute a PR you can do so by just opening it an referring to that "catch-all" issue. Note that some module types may be quite complex (that one should be fine though)

**Reply by CaselIT** [reply]
> For that particular module I don't remember is it was typed in one of the PR yet to merge (I'll check)

seems not, I remembered this one, but it does not add typing to that module https://github.com/sqlalchemy/sqlalchemy/pull/10729

**Reply by jenstroeger** [reply]
Thank you @CaselIT, that makes sense. And with that, I suspect that [sa.Unicode](https://github.com/sqlalchemy/sqlalchemy/blob/a8dbf8763a8fa2ca53cc01033f06681a421bf60b/lib/sqlalchemy/sql/sqltypes.py#L271) also falls into that category
```
error: Call to untyped function "Unicode" in typed context  [no-untyped-call]
```
although, curiously, the other types in the [sa.types](https://docs.sqlalchemy.org/en/20/core/type_basics.html) sub-package seem to work.

**Reply by CaselIT** [reply]
I haven't checked, but from that error it seems likely

**Reply by jenstroeger** [reply]
I’ll continue with the migration for now and I’ll temporarily derive untyped [sa.types](https://docs.sqlalchemy.org/en/20/core/type_basics.html) to a typed sub-class iff they’re missing type annotations, for example for [sa.Unicode](https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.Unicode):
```python
class Unicode(sa.Unicode):
    def __init__(self, length: int | None = None, collation: str | None = None) -> None:
        super().__init__(length=length, collation=collation)  # type: ignore[no-untyped-call]
```
looks like a better solution for now than silencing oodles of warnings across oodles of files 🤔

See also issue https://github.com/sqlalchemy/sqlalchemy/issues/10969 and PR https://github.com/sqlalchemy/sqlalchemy/pull/10970.
